### PR TITLE
fix: [#187870088] poppy formation business address optional

### PR DESCRIPTION
**Pull request by cbrunefearless**
## Description
poppy formation business address optional when submitting to formation

### Ticket

This pull request resolves [#187870088](https://www.pivotaltracker.com/story/show/187870088).

## Code author checklist

- [X] I have rebased this branch from the latest main branch
- [X] I have performed a self-review of my code
- [X] I have created and/or updated relevant documentation (EX: [Engineering Reference/FAQ](https://docs.google.com/document/d/1X4iWSGmBZdHYZ0jGqkwTR3_yyyqI_TiiJptfc8pi9Po)), if necessary
- [X] I have not used any relative imports
- [X] I have checked for and removed instances of unused config from CMS
- [X] I have pruned any instances of unused code
- [X] I have not added any markdown to labels, titles and button text in config
- [X] If I added/updated any values in `userData` (including `profileData`, `formationData` etc), then I added a new migration file
- [X] If I added any new collections to the CMS config, then I updated the search tool and `cmsCollections.ts` (see [CMS Additions in Engineering Reference/FAQ](https://docs.google.com/document/d/1X4iWSGmBZdHYZ0jGqkwTR3_yyyqI_TiiJptfc8pi9Po/edit#heading=h.fu5jdsrcqxbh))
- [X] I have updated relevant `.env` values in both `.env-template` and in Bitwarden

### CODE REVIEW

File: api/src/client/ApiFormationHelpers.ts [hunk 1]
```diff
@@ -212,15 +221,17 @@ export const makePostBody = (
...
-        MainAddress: {
-          Address1: formationFormData.addressLine1,
-          Address2: formationFormData.addressLine2,
-          City: formationFormData.addressMunicipality?.name ?? formationFormData.addressCity ?? "",
-          State: formationFormData.addressState?.name,
-          Province: formationFormData.addressProvince,
-          Zipcode: formationFormData.addressZipCode,
-          Country: formationFormData.addressCountry,
-        },
+        MainAddress: isStartingBusinessAddressPartiallyEmpty()
```
Comment (jphechter) [inline api/src/client/ApiFormationHelpers.ts:224]: [pebble] I'd like this function to be something more generic, and the function itself should clearly state the logic we intend to support inside.

```suggestion
        MainAddress: getBusinessAddress()
```

File: api/src/client/ApiFormationClient.ts [hunk 2]
```diff
@@ -29,11 +29,7 @@ export const ApiFormationClient = (config: ApiConfig, logger: LogWriterType): Fo
     foreignGoodStandingFile: InputFile | undefined
   ): Promise<FormationSubmitResponse> => {
     const postBody = makePostBody(userData, returnUrl, config, foreignGoodStandingFile);
-    const postContentTypePlainOnly =
-      process.env.FEATURE_FORMATION_CONTENT_TYPE_PLAIN_ONLY?.toLowerCase() === "true" ||
```
Comment (FarazA22) [inline api/src/client/ApiFormationClient.ts:33]: are we using the feature flag elsewhere, if not, delete it across the app

File: api/src/client/ApiFormationHelpers.ts [hunk 3]
```diff
@@ -212,15 +221,17 @@ export const makePostBody = (
...
+          ? undefined
+          : {
+              Address1: formationFormData.addressLine1,
+              Address2: formationFormData.addressLine2,
+              City: formationFormData.addressMunicipality?.name ?? formationFormData.addressCity ?? "",
+              State: formationFormData.addressState?.name,
+              Province: formationFormData.addressProvince,
+              Zipcode: formationFormData.addressZipCode,
+              Country: formationFormData.addressCountry,
+            },
```
Comment (jphechter) [inline api/src/client/ApiFormationHelpers.ts:234]: [dust] This works, and we do have other similar logic in this object executed inline, but I think I'd opt for moving this all into a single function.

```suggestion
        MainAddress: getBusinessAddress()
```

File: api/src/client/ApiFormationHelpers.ts [hunk 4]
```diff
@@ -42,6 +42,15 @@ export const makePostBody = (
...
   };
 
+  const isStartingBusinessAddressPartiallyEmpty = (): boolean => {
+    return (
+      currentBusiness.profileData.businessPersona === "STARTING" &&
+      (formationFormData.addressLine1 === "" ||
+        formationFormData.addressMunicipality === undefined ||
+        formationFormData.addressZipCode === "")
+    );
+  };
```
Comment (jphechter) [inline api/src/client/ApiFormationHelpers.ts:52]: [pebble] The function does have a clear name, but given there are 4 independent checks occurring, I think we can make this a little more explicit. If we set the conditions we're concerned about to variables with clear names it makes the logical operation a little clearer.
```suggestion
  const isStartingBusinessAddressPartiallyEmpty = (): boolean => {
   const isStartingBusiness = currentBusiness.profileData.businessPersona === "STARTING"
   const partiallyCompletedAddress = formationFormData.addressLine1 === "" ||
        formationFormData.addressMunicipality === undefined ||
        formationFormData.addressZipCode === ""
    return isStartingBusiness && partiallyCompletedAddress;
  };
```

File: api/src/client/ApiFormationHelpers.ts [hunk 5]
```diff
@@ -42,6 +42,15 @@ export const makePostBody = (
     return currentBusiness.profileData.legalStructureId as FormationLegalType;
   };
 
+  const isStartingBusinessAddressPartiallyEmpty = (): boolean => {
```
Comment (FarazA22) [inline api/src/client/ApiFormationHelpers.ts:45]: requires a test

File: api/src/client/ApiFormationHelpers.ts (hunk 4 above)
Comment (jphechter) [reply api/src/client/ApiFormationHelpers.ts:52]: [boulder] We need to have a test around this logic

File: api/src/client/ApiFormationHelpers.ts (hunk 3 above)
Comment (FarazA22) [reply api/src/client/ApiFormationHelpers.ts:234]: updated

File: api/src/client/ApiFormationHelpers.ts (hunk 4 above)
Comment (FarazA22) [reply api/src/client/ApiFormationHelpers.ts:52]: added tests

File: api/src/client/ApiFormationClient.test.ts [hunk 6]
```diff
@@ -1470,6 +1469,55 @@ describe("ApiFormationClient", () => {
...
+        const userData = generateFormationUserData(
+          { businessPersona: "STARTING" },
+          {},
+          { addressLine1: "", addressMunicipality: undefined, addressZipCode: "" }
+        );
+
+        await client.form(userData, "some-url");
+        const postBody: ApiSubmission = mockAxios.post.mock.calls[0][1] as ApiSubmission;
+        expect(postBody.Formation.BusinessInformation.MainAddress).toBeUndefined();
+      });
```
Comment (jphechter) [inline api/src/client/ApiFormationClient.test.ts:1486]: [sand] Can we add 1 more test to validate this logic doesn't get hit for Foreign businesses?

Comment (cbrunefearless) [reply api/src/client/ApiFormationClient.test.ts:1486]: done

File: api/src/client/ApiFormationHelpers.ts [hunk 7]
```diff
@@ -42,6 +42,26 @@ export const makePostBody = (
     return currentBusiness.profileData.legalStructureId as FormationLegalType;
   };
 
+  const getMainBusinessAddress = (): ApiLocation | undefined => {
+    const isStartingBusinessPartiallyEmpty =
+      currentBusiness.profileData.businessPersona === "STARTING" &&
```
Comment (FarazA22) [inline api/src/client/ApiFormationHelpers.ts:47]: use `isStartingBusiness` instead of `currentBusiness.profileData.businessPersona === "STARTING"`

**Pull request merged**